File: experiments-local/experiment-mu-kimi3-brazil/runs/agent=opencode_language=typescript_model=openrouter/moonshotai/kimi-k3_tooling=none/rep1/src/lib/graph.ts
/**
 * In-memory knowledge graph built over the unified dataset.
 *
 * Node types: team, player, match, competition.
 * Edge types: HOME_IN / AWAY_IN (team -> match), WON / LOST / DREW
 * (team -> match), PLAYS_FOR (player -> team), HAS_NATIONALITY
 * (player -> country label), PLAYED_IN (match -> competition).
 */
import { Competition, Match, Player, Team, matchResult } from "./types.js";
import { Dataset } from "./dataset.js";

export type NodeType = "team" | "player" | "match" | "competition";
export type EdgeType =
  | "HOME_IN"
  | "AWAY_IN"
  | "WON"
  | "LOST"
  | "DREW"
  | "PLAYS_FOR"
  | "HAS_NATIONALITY"
  | "PLAYED_IN";

export interface GraphNode {
  id: string;
  type: NodeType;
  label: string;
}

export interface GraphEdge {
  from: string;
  to: string;
  type: EdgeType;
}

export class KnowledgeGraph {
  readonly nodes = new Map<string, GraphNode>();
  /** Adjacency: node id -> outgoing edges. */
  private outgoing = new Map<string, GraphEdge[]>();
  /** Adjacency: node id -> incoming edges. */
  private incoming = new Map<string, GraphEdge[]>();

  static teamNodeId(team: Team): string {
    return `team:${team.key}`;
  }

  static playerNodeId(player: Player): string {
    return `player:${player.id}`;
  }

  static matchNodeId(match: Match): string {
    return `match:${match.id}`;
  }

  static competitionNodeId(competition: Competition): string {
    return `competition:${competition}`;
  }

  private addNode(node: GraphNode): void {
    if (!this.nodes.has(node.id)) this.nodes.set(node.id, node);
  }

  private addEdge(edge: GraphEdge): void {
    if (!this.outgoing.has(edge.from)) this.outgoing.set(edge.from, []);
    this.outgoing.get(edge.from)!.push(edge);
    if (!this.incoming.has(edge.to)) this.incoming.set(edge.to, []);
    this.incoming.get(edge.to)!.push(edge);
  }

  /** Build the full graph from the loaded dataset. */
  static fromDataset(dataset: Dataset): KnowledgeGraph {
    const g = new KnowledgeGraph();

    for (const competition of Object.values(Competition)) {
      g.addNode({
        id: KnowledgeGraph.competitionNodeId(competition),
        type: "competition",
        label: competition,
      });
    }

    for (const team of dataset.teams.all()) {
      g.addNode({
        id: KnowledgeGraph.teamNodeId(team),
        type: "team",
        label: team.uf ? `${team.name} (${team.uf})` : team.name,
      });
    }

    for (const match of dataset.matches) {
      const matchId = KnowledgeGraph.matchNodeId(match);
      const score =
        match.homeGoals !== null && match.awayGoals !== null
          ? `${match.homeGoals}-${match.awayGoals}`
          : "not played";
      g.addNode({
        id: matchId,
        type: "match",
        label: `${match.date ?? "unknown date"}: ${match.homeTeam.name} ${score} ${match.awayTeam.name}`,
      });
      g.addEdge({ from: matchId, to: KnowledgeGraph.competitionNodeId(match.competition), type: "PLAYED_IN" });

      const homeId = KnowledgeGraph.teamNodeId(match.homeTeam);
      const awayId = KnowledgeGraph.teamNodeId(match.awayTeam);
      g.addEdge({ from: homeId, to: matchId, type: "HOME_IN" });
      g.addEdge({ from: awayId, to: matchId, type: "AWAY_IN" });

      const result = matchResult(match);
      if (result === "home") {
        g.addEdge({ from: homeId, to: matchId, type: "WON" });
        g.addEdge({ from: awayId, to: matchId, type: "LOST" });
      } else if (result === "away") {
        g.addEdge({ from: awayId, to: matchId, type: "WON" });
        g.addEdge({ from: homeId, to: matchId, type: "LOST" });
      } else if (result === "draw") {
        g.addEdge({ from: homeId, to: matchId, type: "DREW" });
        g.addEdge({ from: awayId, to: matchId, type: "DREW" });
      }
    }

    for (const player of dataset.players) {
      const playerId = KnowledgeGraph.playerNodeId(player);
      g.addNode({ id: playerId, type: "player", label: player.name });
      if (player.teamKey) {
        const team = dataset.teams.get(player.teamKey);
        if (team) {
          g.addEdge({ from: playerId, to: KnowledgeGraph.teamNodeId(team), type: "PLAYS_FOR" });
        }
      }
    }

    return g;
  }

  /** Neighbors of a node (both directions), optionally filtered by edge type. */
  neighbors(nodeId: string, edgeType?: EdgeType): { edge: GraphEdge; node: GraphNode }[] {
    const out: { edge: GraphEdge; node: GraphNode }[] = [];
    for (const e of this.outgoing.get(nodeId) ?? []) {
      if (edgeType && e.type !== edgeType) continue;
      const n = this.nodes.get(e.to);
      if (n) out.push({ edge: e, node: n });
    }
    for (const e of this.incoming.get(nodeId) ?? []) {
      if (edgeType && e.type !== edgeType) continue;
      const n = this.nodes.get(e.from);
      if (n) out.push({ edge: e, node: n });
    }
    return out;
  }

  degree(nodeId: string): number {
    return (this.outgoing.get(nodeId)?.length ?? 0) + (this.incoming.get(nodeId)?.length ?? 0);
  }
}

File: experiments-local/experiment-mu-glm52-claudecode/runs/agent=claude-code_language=python_model=z-ai/glm-5.2_tooling=none/rep2/brazilian_soccer_mcp/graph.py
"""In-memory knowledge graph for Brazilian soccer data.

Context
-------
TASK.md asks for an MCP server that provides a *knowledge graph interface* for
Brazilian soccer.  Rather than require an external graph database (e.g. Neo4j),
this module implements a small, dependency-free in-memory property graph that is
built once at load time from the CSV datasets.

Nodes are typed entities (``Team``, ``Player``, ``Match``, ``Competition``,
``Season``) and edges describe the relationships between them
(``PLAYED_IN``, ``PARTICIPATED_IN``, ``HELD_IN``, ``WON``, ``DREW``, ``LOST``).
The query layer in ``queries.py`` walks this graph (and the parallel record
lists) to answer natural-language questions.

The graph is the conceptual API; record lists remain the fast path for
aggregations, but the graph gives the "knowledge graph interface" the spec
calls for and lets us answer relationship queries such as
"What competitions has Palmeiras played in?".
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List, Optional, Tuple

from .normalize import team_key


@dataclass
class Node:
    """A node in the knowledge graph."""

    id: str
    type: str  # team | player | match | competition | season
    label: str
    properties: Dict[str, Any] = field(default_factory=dict)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Node) and self.id == other.id and self.type == other.type

    def __hash__(self) -> int:
        return hash((self.type, self.id))


@dataclass
class Edge:
    """A directed edge between two nodes."""

    source: str  # node id
    target: str  # node id
    relation: str  # e.g. PLAYED_IN, WON, LOST, DREW, PARTICIPATED_IN, HELD_IN
    properties: Dict[str, Any] = field(default_factory=dict)


class KnowledgeGraph:
    """A simple property graph over teams, players, matches, competitions."""

    def __init__(self) -> None:
        self.nodes: Dict[str, Node] = {}
        self.edges: List[Edge] = []
        # Adjacency: source -> [(relation, target, properties)]
        self._adj: Dict[str, List[Edge]] = defaultdict(list)
        # Reverse adjacency: target -> [edge, ...]
        self._radj: Dict[str, List[Edge]] = defaultdict(list)
        # Indexes by type and by match key for teams.
        self._by_type: Dict[str, List[Node]] = defaultdict(list)
        self._team_by_key: Dict[str, str] = {}

    # -- node / edge mutation ------------------------------------------------

    def add_node(self, type_: str, id_: str, label: str, **properties) -> Node:
        """Add (or upsert) a node and return it."""
        node = Node(id=id_, type=type_, label=label, properties=properties)
        if id_ not in self.nodes:
            self.nodes[id_] = node
            self._by_type[type_].append(node)
            if type_ == "team":
                self._team_by_key[id_] = id_
        else:
            # Merge properties.
            self.nodes[id_].properties.update(properties)
        return self.nodes[id_]

    def add_edge(self, source: str, target: str, relation: str, **properties) -> Edge:
        """Add a directed edge *source* -> *target* with *relation*."""
        edge = Edge(source=source, target=target, relation=relation, properties=properties)
        self.edges.append(edge)
        self._adj[source].append(edge)
        self._radj[target].append(edge)
        return edge

    # -- lookups -------------------------------------------------------------

    def get_node(self, id_: str) -> Optional[Node]:
        return self.nodes.get(id_)

    def nodes_by_type(self, type_: str) -> List[Node]:
        return list(self._by_type.get(type_, []))

    def team_node_id(self, name: str) -> str:
        """Return the graph node id for a team name (accent/case-insensitive)."""
        return team_key(name)

    def team_label(self, name: str) -> str:
        """Return the preferred display label for a team name."""
        node = self.nodes.get(team_key(name))
        if node is not None:
            return node.label
        from .normalize import normalize_team

        return normalize_team(name)

    def find_team(self, query: str) -> Optional[Node]:
        """Find a team node by accent/case-insensitive substring match."""
        q = team_key(query)
        if not q:
            return None
        # Exact key match first.
        if q in self._team_by_key:
            return self.nodes[self._team_by_key[q]]
        # Substring match on the canonical key.
        matches = [n for n in self._by_type.get("team", []) if q in n.id]
        if not matches:
            # Try matching on the stripped label too.
            matches = [
                n for n in self._by_type.get("team", [])
                if q in n.label.lower() or q in n.id
            ]
        return matches[0] if matches else None

    def neighbors(self, source: str, relation: Optional[str] = None) -> List[Edge]:
        """Return outgoing edges from *source*, optionally filtered by relation."""
        edges = self._adj.get(source, [])
        if relation is None:
            return list(edges)
        return [e for e in edges if e.relation == relation]

    def incoming(self, target: str, relation: Optional[str] = None) -> List[Edge]:
        """Return incoming edges to *target*, optionally filtered by relation."""
        edges = self._radj.get(target, [])
        if relation is None:
            return list(edges)
        return [e for e in edges if e.relation == relation]

    def relations_of(self, source_id: str, relation: str) -> List[Node]:
        """Return the nodes that *source_id* connects to via *relation*."""
        return [self.nodes[e.target] for e in self._adj.get(source_id, []) if e.relation == relation]

    # -- summary -------------------------------------------------------------

    def summary(self) -> Dict[str, int]:
        """Return counts of nodes/edges by type/relation."""
        counts: Dict[str, int] = defaultdict(int)
        for n in self.nodes.values():
            counts[f"nodes/{n.type}"] += 1
        for e in self.edges:
            counts[f"edges/{e.relation}"] += 1
        counts["edges/total"] = len(self.edges)
        return dict(counts)
